README.md
[![NPM version](https://badge.fury.io/js/%40cdk-constructs-zone%2Fsuper-ec2.svg)](https://badge.fury.io/js/%40cdk-constructs-zone%2Fsuper-ec2)
[![PyPI version](https://badge.fury.io/py/super-ec2.svg)](https://badge.fury.io/py/super-ec2)
![Release](https://github.com/cdk-constructs-zone/super-ec2/workflows/Release/badge.svg)

![Downloads](https://img.shields.io/badge/-DOWNLOADS:-brightgreen?color=gray)
![npm](https://img.shields.io/npm/dt/@cdk-constructs-zone/super-ec2?label=npm&color=orange)
![PyPI](https://img.shields.io/pypi/dm/super-ec2?label=pypi&color=blue)

![](https://img.shields.io/badge/jenkins-ec2-green=?style=plastic&logo=appveyor)

# Welcome to `@cdk-constructs-zone/super-ec2`

This repository template helps you create EC2 .

## Sample

### Jenkins
* Simplest deployment: It would creat a VPC and ALB by default.
```ts
import * as cdk from '@aws-cdk/core';
import { JenkinsEC2, ELBtype } from '@cdk-constructs-zone/super-ec2';

const app = new cdk.App();

const stack = new cdk.Stack(app, 'demo');

const jks = new JenkinsEC2(stack, 'superJks', {});

new cdk.CfnOutput(stack, 'loadbalancerDNS', {
  value: jks.loadbalancer.loadBalancerDnsName,
});
new cdk.CfnOutput(stack, 'connect-to-instance', {
  value: `aws ssm start-session --target ${jks.instance.instanceId}`,
});
```
* Deploy Jenkins with self-defined VPC and NLB
```ts
const jks = new JenkinsEC2(stack, 'superJks', {
  vpc: Vpc.fromLookup(stack, 'defaultVPC', { isDefault: true }),
  loadbalancerType: ELBtype.NLB,
});
```
* Deploy Jenkins with R53 records: If `acm` is not given, it would create a certificate validated from DNS by default. 
```ts
const jks = new JenkinsEC2(stack, 'superJks', {
  vpc: Vpc.fromLookup(stack, 'defaultVPC', { isDefault: true }),
  loadbalancerType: ELBtype.NLB,
  domain: {
    acm: Certificate.fromCertificateArn(stack, 'cert',
      'arn:aws:xxx',
    ),
    hostedZoneId: 'xxx',
    zoneName: 'bbb.ccc',
    recordName: 'aaa',
  },
});
```

Note: Jenkins initial admin password has been written to `/home/ec2-user/jenkins-data/secrets/initialAdminPassword`. You can access EC2 instance using [ssm tool](https://docs.aws.amazon.com/systems-manager/latest/userguide/session-manager-working-with-sessions-start.html).
```
aws ssm start-session --target instance-id
```

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.23](https://github.com/cdk-constructs-zone/super-ec2/compare/v0.0.22...v0.0.23) (2021-04-21)

### [0.0.22](https://github.com/cdk-constructs-zone/super-ec2/compare/v0.0.21...v0.0.22) (2021-04-20)

### [0.0.21](https://github.com/cdk-constructs-zone/super-ec2/compare/v0.0.20...v0.0.21) (2021-04-19)

### [0.0.20](https://github.com/cdk-constructs-zone/super-ec2/compare/v0.0.19...v0.0.20) (2021-04-18)

### [0.0.19](https://github.com/cdk-constructs-zone/super-ec2/compare/v0.0.18...v0.0.19) (2021-04-18)

### [0.0.18](https://github.com/cdk-constructs-zone/super-ec2/compare/v0.0.17...v0.0.18) (2021-04-17)

### [0.0.17](https://github.com/cdk-constructs-zone/super-ec2/compare/v0.0.16...v0.0.17) (2021-04-16)


### Features

* add elb and route53 option for JenkinsEC2 ([#13](https://github.com/cdk-constructs-zone/super-ec2/issues/13)) ([10d444d](https://github.com/cdk-constructs-zone/super-ec2/commit/10d444dde8dd7ce2099568c312dbfb9030ce512e))

### [0.0.16](https://github.com/cdk-constructs-zone/super-ec2/compare/v0.0.15...v0.0.16) (2021-04-16)

### [0.0.15](https://github.com/cdk-constructs-zone/super-ec2/compare/v0.0.14...v0.0.15) (2021-04-15)

### [0.0.14](https://github.com/cdk-constructs-zone/super-ec2/compare/v0.0.13...v0.0.14) (2021-04-14)

### [0.0.13](https://github.com/cdk-constructs-zone/super-ec2/compare/v0.0.12...v0.0.13) (2021-04-13)

### [0.0.12](https://github.com/cdk-constructs-zone/super-ec2/compare/v0.0.11...v0.0.12) (2021-04-12)

### [0.0.11](https://github.com/cdk-constructs-zone/super-ec2/compare/v0.0.10...v0.0.11) (2021-04-11)

### [0.0.10](https://github.com/cdk-constructs-zone/super-ec2/compare/v0.0.9...v0.0.10) (2021-04-07)

### [0.0.9](https://github.com/cdk-constructs-zone/super-ec2/compare/v0.0.8...v0.0.9) (2021-04-06)

### [0.0.8](https://github.com/cdk-constructs-zone/super-ec2/compare/v0.0.7...v0.0.8) (2021-04-05)

### [0.0.7](https://github.com/cdk-constructs-zone/super-ec2/compare/v0.0.6...v0.0.7) (2021-04-02)


### Features

* **doc:** Update issue templates ([67e2caf](https://github.com/cdk-constructs-zone/super-ec2/commit/67e2caf957c355324fe95ff5ae2c57d44b4cd98e))

### [0.0.6](https://github.com/cdk-constructs-zone/super-ec2/compare/v0.0.5...v0.0.6) (2021-03-30)


### Features

* **doc:** Update issue templates ([67e2caf](https://github.com/cdk-constructs-zone/super-ec2/commit/67e2caf957c355324fe95ff5ae2c57d44b4cd98e))

### [0.0.5](https://github.com/cdk-constructs-zone/super-ec2/compare/v0.0.4...v0.0.5) (2021-03-30)

### [0.0.4](https://github.com/cdk-constructs-zone/super-ec2/compare/v0.0.3...v0.0.4) (2021-03-30)

### [0.0.3](https://github.com/cdk-constructs-zone/super-ec2/compare/v0.0.2...v0.0.3) (2021-03-30)

### [0.0.2](https://github.com/guan840912/super-ec2/compare/v0.0.1...v0.0.2) (2021-03-29)

### 0.0.1 (2021-03-29)


### Features

* jenkins base ([0d86273](https://github.com/guan840912/super-ec2/commit/0d86273ddd8d1d58ff23cfc9d6c75b52091ac948))


test/super-ec2.test.ts
import { App, Stack } from '@aws-cdk/core';
import { ELBtype, JenkinsEC2 } from '../src';
import '@aws-cdk/assert/jest';

describe('SpotInstance', () => {
  let mockApp: App;
  let stack: Stack;

  beforeEach(() => {
    mockApp = new App();
    stack = new Stack(mockApp, 'testing-stack');
  });

  test('create single spot instance', () => {
    new JenkinsEC2(stack, 'JenkinsEC2', {});
    expect(stack).toHaveResourceLike('AWS::EC2::Instance');
  });
});

describe('CreateJksALB', () => {
  let mockApp: App;
  let stack: Stack;

  beforeEach(() => {
    mockApp = new App();
    stack = new Stack(mockApp, 'testing-stack');
  });

  test('create jks with alb', () => {
    new JenkinsEC2(stack, 'JenkinsEC2', { loadbalancerType: ELBtype.ALB });
    expect(stack).toHaveResource('AWS::ElasticLoadBalancingV2::LoadBalancer', {
      Type: 'application',
    });
    expect(stack).toHaveResource('AWS::ElasticLoadBalancingV2::TargetGroup', {
      Port: 8080,
      Protocol: 'HTTP',
      TargetType: 'instance',
    });
    expect(stack).toHaveResource('AWS::ElasticLoadBalancingV2::Listener', {
      Port: 80,
      Protocol: 'HTTP',
    });
  });
});

describe('CreateJksNLB', () => {
  let mockApp: App;
  let stack: Stack;

  beforeEach(() => {
    mockApp = new App();
    stack = new Stack(mockApp, 'testing-stack');
  });

  test('create jks with nlb', () => {
    new JenkinsEC2(stack, 'JenkinsEC2', { loadbalancerType: ELBtype.NLB });
    expect(stack).toHaveResource('AWS::ElasticLoadBalancingV2::LoadBalancer', {
      Type: 'network',
    });
    expect(stack).toHaveResource('AWS::ElasticLoadBalancingV2::TargetGroup', {
      Port: 8080,
      Protocol: 'TCP',
      TargetType: 'instance',
    });
    expect(stack).toHaveResource('AWS::ElasticLoadBalancingV2::Listener', {
      Port: 80,
      Protocol: 'TCP',
    });
  });
});


describe('CreateJksWithALBDomain', () => {
  let mockApp: App;
  let stack: Stack;

  beforeEach(() => {
    mockApp = new App();
    stack = new Stack(mockApp, 'testing-stack');
  });

  test('create jks with alb and domain', () => {
    new JenkinsEC2(stack, 'JenkinsEC2', {
      loadbalancerType: ELBtype.ALB,
      domain: {
        hostedZoneId: 'aaabbb',
        zoneName: 'aaa.bbb',
        recordName: 'superjks',
      },
    });
    expect(stack).toHaveResource('AWS::ElasticLoadBalancingV2::LoadBalancer', {
      Type: 'application',
    });
    expect(stack).toHaveResource('AWS::ElasticLoadBalancingV2::Listener', {
      Port: 80,
      Protocol: 'HTTP',
      DefaultActions: [
        {
          RedirectConfig: {
            Host: '#{host}',
            Path: '/#{path}',
            Port: '443',
            Protocol: 'HTTPS',
            Query: '/#{query}',
            StatusCode: 'HTTP_302',
          },
          Type: 'redirect',
        },
      ],
    });
    expect(stack).toHaveResource('AWS::ElasticLoadBalancingV2::Listener', {
      Port: 443,
      Protocol: 'HTTPS',
    });
    expect(stack).toHaveResource('AWS::ElasticLoadBalancingV2::TargetGroup', {
      Port: 8080,
      Protocol: 'HTTP',
      TargetType: 'instance',
    });
    expect(stack).toHaveResource('AWS::Route53::RecordSet', {
      Name: 'superjks.aaa.bbb.',
      HostedZoneId: 'aaabbb',
    });
    expect(stack).toHaveResource('AWS::CertificateManager::Certificate', {
      DomainName: '*.aaa.bbb',
      DomainValidationOptions: [
        {
          DomainName: '*.aaa.bbb',
          HostedZoneId: 'aaabbb',
        },
      ],
      ValidationMethod: 'DNS',
    });
  });
});

describe('CreateJksWithNLBDomain', () => {
  let mockApp: App;
  let stack: Stack;

  beforeEach(() => {
    mockApp = new App();
    stack = new Stack(mockApp, 'testing-stack');
  });

  test('create jks with nlb and domain', () => {
    new JenkinsEC2(stack, 'JenkinsEC2', {
      loadbalancerType: ELBtype.NLB,
      domain: {
        hostedZoneId: 'aaabbb',
        zoneName: 'aaa.bbb',
        recordName: 'superjks',
      },
    });
    expect(stack).toHaveResource('AWS::ElasticLoadBalancingV2::LoadBalancer', {
      Type: 'network',
    });
    expect(stack).toHaveResource('AWS::ElasticLoadBalancingV2::Listener', {
      Port: 443,
      Protocol: 'TLS',
    });
    expect(stack).toHaveResource('AWS::ElasticLoadBalancingV2::TargetGroup', {
      Port: 8080,
      Protocol: 'TCP',
      TargetType: 'instance',
    });
    expect(stack).toHaveResource('AWS::Route53::RecordSet', {
      Name: 'superjks.aaa.bbb.',
      HostedZoneId: 'aaabbb',
    });
    expect(stack).toHaveResource('AWS::CertificateManager::Certificate', {
      DomainName: '*.aaa.bbb',
      DomainValidationOptions: [
        {
          DomainName: '*.aaa.bbb',
          HostedZoneId: 'aaabbb',
        },
      ],
      ValidationMethod: 'DNS',
    });
  });
});

src/ami.ts
import * as ec2 from '@aws-cdk/aws-ec2';

export enum AmiOSType {
  /**
   * Ubuntu 18.04 ami.
   */
  UBUNTU_18_04 = 'UBUNTU_18_04',

  /**
   * Ubuntu 20.04 ami.
   */
  UBUNTU_20_04 = 'UBUNTU_20_04',

  /**
   * CentOS 7 ami.
   */
  CENTOS_7 = 'CENTOS_7',

  /**
   * CentOS 8 ami.
   */
  CENTOS_8 = 'CENTOS_8',

  /**
   * Amazon Linux 2 ami.
   */
  AMAZON_LINUX_2 = 'AMAZON_LINUX_2',

  /**
   * Amazon Linux  ami.
   */
  AMAZON_LINUX = 'AMAZON_LINUX',

}

export function amiFinder(amiostype: AmiOSType ): ec2.IMachineImage {
  switch (amiostype) {
    case AmiOSType.CENTOS_7:
      return ec2.MachineImage.lookup({
        filters: {
          ['product-code']: ['aw0evgkw8e5c1q413zgy5pjce'],
        },
        name: '*CentOS*',
        owners: ['aws-marketplace'],
      });
    case AmiOSType.CENTOS_8:
      return ec2.MachineImage.lookup({
        filters: {
          ['product-code']: ['aw0evgkw8e5c1q413zgy5pjce'],
        },
        name: '*CentOS*',
        owners: ['aws-marketplace'],
      });
    case AmiOSType.UBUNTU_18_04:
      return ec2.MachineImage.lookup({
        //name: '*ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server*',
        name: '*ubuntu-bionic-18.04-amd64-*',
        owners: ['099720109477'],
        filters: {
          ['root-device-type']: ['ebs'],
        },
      });
    case AmiOSType.UBUNTU_20_04:
      return ec2.MachineImage.lookup({
        name: '*ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server*',
        owners: ['099720109477'],
        filters: {
          ['root-device-type']: ['ebs'],
        },
      });
    case AmiOSType.AMAZON_LINUX:
      return ec2.MachineImage.latestAmazonLinux();
    case AmiOSType.AMAZON_LINUX_2:
      return ec2.MachineImage.latestAmazonLinux( { generation: ec2.AmazonLinuxGeneration.AMAZON_LINUX_2 });
  }
}

.projenrc.js
const { AwsCdkConstructLibrary } = require('projen');
const { Automation } = require('projen-automate-it');

const PROJECT_NAME = 'super-ec2';
const PROJECT_DESCRIPTION = 'A construct lib for AWS CDK EC2';
const AUTOMATION_TOKEN = 'AUTOMATION_GITHUB_TOKEN';

const project = new AwsCdkConstructLibrary({
  name: PROJECT_NAME,
  description: PROJECT_DESCRIPTION,
  author: '@cdk-constructs-zone',
  authorAddress: 'https://github.com/cdk-constructs-zone',
  authorOrganization: true,
  authorUrl: 'https://github.com/cdk-constructs-zone',
  npmAccess: 'public',
  cdkVersion: '1.96.0',
  defaultReleaseBranch: 'main',
  jsiiFqn: 'projen.AwsCdkConstructLibrary',
  name: '@cdk-constructs-zone/super-ec2',
  repositoryUrl: 'https://github.com/cdk-constructs-zone/super-ec2.git',
  keywords: ['aws', 'cdk', 'ec2', 'construct'],
  dependabot: false,
  catalog: {
    twitter: 'neil_kuan',
    announce: false,
  },
  compat: true,
  stability: 'experimental',
  cdkDependencies: [
    '@aws-cdk/aws-elasticloadbalancingv2-targets',
    '@aws-cdk/aws-elasticloadbalancingv2',
    '@aws-cdk/aws-certificatemanager',
    '@aws-cdk/aws-route53',
    '@aws-cdk/aws-route53-targets',
    '@aws-cdk/aws-iam',
    '@aws-cdk/aws-ec2',
    '@aws-cdk/aws-autoscaling',
    '@aws-cdk/aws-autoscaling-hooktargets',
    '@aws-cdk/aws-lambda',
    '@aws-cdk/aws-logs',
    '@aws-cdk/aws-s3',
    '@aws-cdk/aws-s3-assets',
    '@aws-cdk/core',
    '@aws-cdk/aws-sns',
    '@aws-cdk/aws-sns-subscriptions',
    '@aws-cdk/custom-resources',
  ],
  devDeps: [
    'xmldom',
    'projen-automate-it',
  ],
  bundledDeps: ['projen-automate-it'],
  python: {
    distName: 'super-ec2',
    module: 'super_ec2',
  },
});
const automation = new Automation(project, {
  automationToken: AUTOMATION_TOKEN,
});
automation.projenYarnUpgrade();

const common_exclude = ['cdk.out', 'cdk.context.json', 'images', 'yarn-error.log', '*.zip'];
project.gitignore.exclude(...common_exclude);
project.npmignore.exclude(...common_exclude, 'images');
project.synth();


src/super-ec2.ts
import * as ec2 from '@aws-cdk/aws-ec2';
import * as iam from '@aws-cdk/aws-iam';
import * as cdk from '@aws-cdk/core';
import { AmiOSType, amiFinder } from './ami';
export interface ISuperEC2BaseProps {
  /**
   * Super EC2 OS you want.
   *
   * @default - Amzaon Linux 2.
   */
  amiOSType?: AmiOSType;

  /**
   * Super EC2 Vpc.
   *
   * @default - Create a new Vpc.
   */
  vpc?: ec2.IVpc;

  /**
   * Super EC2 Instance Type.
   *
   * @default - t3.small.
   */
  instanceType?: ec2.InstanceType;

}

export abstract class SuperEC2Base extends cdk.Construct {
  readonly instance: ec2.Instance;
  readonly vpc: ec2.IVpc;
  readonly userData: ec2.UserData;
  readonly defaultSecurityGroup: ec2.SecurityGroup;
  constructor(scope: cdk.Construct, id: string, props: ISuperEC2BaseProps ) {
    super(scope, id);
    this.vpc = props.vpc ?? new ec2.Vpc(this, 'SuperEC2Vpc', {
      enableDnsHostnames: true,
      enableDnsSupport: true,
      natGateways: 1,
    });
    this.userData = ec2.UserData.forLinux();
    this.defaultSecurityGroup = new ec2.SecurityGroup(this, 'defaultSecurityGroup', {
      vpc: this.vpc,
    });
    this.instance = new ec2.Instance(this, 'SuperEC2', {
      machineImage: props.amiOSType ? amiFinder(props?.amiOSType) : amiFinder(AmiOSType.AMAZON_LINUX_2),
      vpc: this.vpc,
      instanceType: props.instanceType ?? new ec2.InstanceType('t3.small'),
      userData: this.userData,
      securityGroup: this.defaultSecurityGroup,
    });
    this.instance.role.addManagedPolicy(iam.ManagedPolicy.fromAwsManagedPolicyName('AmazonSSMManagedInstanceCore'));
  }
}

src/super-ec2-interfaces.ts
import * as certmgr from '@aws-cdk/aws-certificatemanager';
import * as elbv2 from '@aws-cdk/aws-elasticloadbalancingv2';
import * as r53 from '@aws-cdk/aws-route53';
import * as r53tg from '@aws-cdk/aws-route53-targets';
import * as cdk from '@aws-cdk/core';

export enum ELBtype {

  /**
   * Application Load Balancer
   */
  ALB = 'ALB',

  /**
   * Network Load Balancer
   */
  NLB = 'NLB',

}

export interface IDomainProps {
  /**
   * Provide a certificate
   * @default - Create a new certificate (validate from DNS)
   */
  acm?: certmgr.ICertificate;
  /**
   * HostZoneID
   */
  hostedZoneId: string;
  /**
   * zonename (e.g., `cdk-construct-zone.com`)
   */
  zoneName: string;
  /**
   * recordname (e.g., superjks)
   */
  recordName: string;
  /**
   * record cache time
   * @deafult - 5 mins.
   */
  ttl?: cdk.Duration;

}

export interface ISuperDomainProps extends IDomainProps{
  loadbalancer: elbv2.ApplicationLoadBalancer | elbv2.NetworkLoadBalancer;
}

export class SuperDomain extends cdk.Construct {
  readonly acm: certmgr.ICertificate;
  readonly record: r53.ARecord;
  readonly zone: r53.IHostedZone;
  constructor(scope: cdk.Construct, id: string, props: ISuperDomainProps) {
    super(scope, id);
    this.zone = this._setzone(props);
    this.acm = props.acm ?? this._setacm(props);
    this.record = this._setaliasrecord(props);
  }
  private _setzone(props: ISuperDomainProps) {
    return r53.HostedZone.fromHostedZoneAttributes(this, 'zone', {
      hostedZoneId: props.hostedZoneId,
      zoneName: props.zoneName,
    });
  }
  private _setacm(props: ISuperDomainProps) {
    return new certmgr.Certificate(this, 'cert', {
      domainName: `*.${props.zoneName}`,
      validation: certmgr.CertificateValidation.fromDns(this.zone),
    });
  }
  private _setaliasrecord(props: ISuperDomainProps ) {
    return new r53.ARecord(this, 'alias-alb', {
      zone: this.zone,
      target: r53.RecordTarget.fromAlias(new r53tg.LoadBalancerTarget(props.loadbalancer)),
      recordName: props.recordName,
      ttl: props.ttl ?? cdk.Duration.minutes(5),
    });
  }
}

src/jenkins-ec2.ts
import * as certmgr from '@aws-cdk/aws-certificatemanager';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as elbv2 from '@aws-cdk/aws-elasticloadbalancingv2';
import * as targets from '@aws-cdk/aws-elasticloadbalancingv2-targets';
import * as cdk from '@aws-cdk/core';
import { SuperEC2Base, ISuperEC2BaseProps } from './super-ec2';
import { ELBtype, IDomainProps, SuperDomain } from './super-ec2-interfaces';

export interface IJenkinsEC2Props extends ISuperEC2BaseProps {
  /**
   * Provide a loadbalancer.
   *
   * Only support ALB and NLB.
   * @default - Create ApplicationLoadBalancer
   */
  loadbalancer?: elbv2.ApplicationLoadBalancer | elbv2.NetworkLoadBalancer;
  /**
   * ELB type.
   * @default - ELBtype.ALB
   */
  loadbalancerType?: ELBtype;
  /**
   * Provide domain attribute.
   * @default - Not use certificate and route53
   */
  domain?: IDomainProps;
};


export class JenkinsEC2 extends SuperEC2Base {
  readonly loadbalancer: elbv2.ApplicationLoadBalancer | elbv2.NetworkLoadBalancer;
  readonly domain?: SuperDomain;

  constructor(scope: cdk.Construct, id: string, props: IJenkinsEC2Props) {
    super(scope, id, props);
    this.userData.addCommands(...this.jenkinsUserData());
    this._addIngressRule(this.defaultSecurityGroup, [22, 8080, 50000]);
    this.loadbalancer = props.loadbalancer?? this._setloadbalancer(props.loadbalancerType);

    if (props.domain) {
      this.domain = new SuperDomain(this, 'superDomain',
        { ...props.domain, loadbalancer: this.loadbalancer },
      );
      this._setlistener(this.domain.acm);
    } else {
      this._set80listener();
    }
  }
  public jenkinsUserData(): string[] {
    return ['yum install docker -y',
      'systemctl start docker',
      'usermod -aG docker ec2-user',
      'usermod -aG docker ssm-user',
      'chmod +x /var/run/docker.sock',
      'systemctl restart docker && systemctl enable docker',
      'mkdir /home/ec2-user/jenkins-data',
      'docker run --name jks --rm -d -u root -p 8080:8080 -p 50000:50000 -v /home/ec2-user/jenkins-data:/var/jenkins_home -v /var/run/docker.sock:/var/run/docker.sock -v /home/ec2-user:/home jenkinsci/blueocean'];
  };
  private _addIngressRule(SecurityGroup: ec2.SecurityGroup, Ports: number[]): ec2.SecurityGroup {
    Ports.forEach(port => {
      SecurityGroup.addIngressRule(ec2.Peer.anyIpv4(), ec2.Port.tcp(port));
    });
    return SecurityGroup;
  }
  private _setloadbalancer(loadbalancerType: ELBtype | undefined): elbv2.ApplicationLoadBalancer | elbv2.NetworkLoadBalancer {
    switch (loadbalancerType) {
      case ELBtype.ALB:
        return new elbv2.ApplicationLoadBalancer(this, 'jksALB', {
          vpc: this.vpc,
          loadBalancerName: 'jksALB',
          internetFacing: true,
        });
      case ELBtype.NLB:
        return new elbv2.NetworkLoadBalancer(this, 'jksNLB', {
          vpc: this.vpc,
          loadBalancerName: 'jksNLB',
          internetFacing: true,
        });
      default: {
        return new elbv2.ApplicationLoadBalancer(this, 'jksALB', {
          vpc: this.vpc,
          loadBalancerName: 'jksALB',
          internetFacing: true,
        });
      }
    }
  }
  private _setlistener(acm: certmgr.ICertificate) {
    const httpslistener = this.loadbalancer.addListener('httpsListener', {
      certificates: [elbv2.ListenerCertificate.fromArn(acm.certificateArn)],
      port: 443,
    });
    if (this.loadbalancer instanceof elbv2.ApplicationLoadBalancer) {
      this.loadbalancer.addListener('httpListener', {
        port: 80,
        open: true,
        defaultAction: elbv2.ListenerAction.redirect({
          protocol: 'HTTPS',
          host: '#{host}',
          path: '/#{path}',
          query: '/#{query}',
          port: '443',
        }),
      });
    }
    if (httpslistener instanceof elbv2.ApplicationListener) {
      httpslistener.addTargets('Targets', {
        port: 8080,
        protocol: elbv2.ApplicationProtocol.HTTP,
        targets: [new targets.InstanceTarget(this.instance)],
      });
      httpslistener.connections.allowTo(this.instance, ec2.Port.tcp(8080));
    } else if (httpslistener instanceof elbv2.NetworkListener) {
      httpslistener.addTargets('Targets', {
        port: 8080,
        protocol: elbv2.Protocol.TCP,
        targets: [new targets.InstanceTarget(this.instance)],
      });
    }
  }
  private _set80listener() {
    const httplistener = this.loadbalancer.addListener('httpListener', {
      port: 80,
    });
    if (httplistener instanceof elbv2.ApplicationListener) {
      httplistener.addTargets('Targets', {
        port: 8080,
        protocol: elbv2.ApplicationProtocol.HTTP,
        targets: [new targets.InstanceTarget(this.instance)],
      });
      httplistener.connections.allowTo(this.instance, ec2.Port.tcp(8080));
    } else if (httplistener instanceof elbv2.NetworkListener) {
      httplistener.addTargets('Targets', {
        port: 8080,
        protocol: elbv2.Protocol.TCP,
        targets: [new targets.InstanceTarget(this.instance)],
      });
    }
  }
}
